asynciterator.ts

```diff
@@ -1621,145 +1621,124 @@ export class MultiTransformIterator<S, D = S> extends TransformIterator<S, D> {
-  private _currentSource = -1;
-  protected _destroySources: boolean;
+export class UnionIterator<T> extends AsyncIterator<T> {
+  private _sources : AsyncIterator<AsyncIterator<T>>;
+  private _buffer = new LinkedList<AsyncIterator<T>>();
+  private _sourceStarted: boolean = false;
+  private _destroySources: boolean;
+  private _maxParallelIterators: number;
 
   /**
     Creates a new `UnionIterator`.
-    @param {module:asynciterator.AsyncIterator|Array} [sources] The sources to read from
+    @param {module:asynciterator.AsyncIterator|Array} [sources] The sources to read from.
     @param {object} [options] Settings of the iterator
     @param {boolean} [options.destroySource=true] Whether the sources should be destroyed when transformed iterator is closed or destroyed
+    @param {number} [options.maxParallelIterators=Infinity] The maximum number of iterators that can be read from simultaneously.
+    Setting equal to 1 will cause iterators to be read from sequentially and in order.
   */
   constructor(sources: AsyncIteratorOrArray<AsyncIterator<T>> |
                        AsyncIteratorOrArray<Promise<AsyncIterator<T>>> |
-                       AsyncIteratorOrArray<MaybePromise<AsyncIterator<T>>>,
-              options: BufferedIteratorOptions & { destroySources?: boolean } = {}) {
-    super(options);
-    const autoStart = options.autoStart !== false;
+                       AsyncIteratorOrArray<MaybePromise<AsyncIterator<T>>> |
+                       MaybePromise<AsyncIteratorOrArray<MaybePromise<IterableSource<T>>>>,
+              options: { destroySources?: boolean, maxParallelIterators?: number } = {}) {
+    super();
+    this._addSource(this._sources = wrap(sources).map<AsyncIterator<T>>(wrap));
 
-    // Sources have been passed as an iterator
-    if (isEventEmitter(sources)) {
-      sources.on('error', error => this.emit('error', error));
-      this._pending = { loading: false, sources: sources as AsyncIterator<MaybePromise<AsyncIterator<T>>> };
-      if (autoStart)
-        this._loadSources();
-    }
-    // Sources have been passed as a non-empty array
-    else if (Array.isArray(sources) && sources.length > 0) {
-      for (const source of sources)
-        this._addSource(source as MaybePromise<InternalSource<T>>);
-    }
-    // Sources are an empty list
-    else if (autoStart) {
-      this.close();
-    }
     // Set other options
     this._destroySources = options.destroySources !== false;
+    this._maxParallelIterators = options.maxParallelIterators || Infinity;
+
+    this.readable = this._sources.readable;
   }
 
-  // Loads pending sources into the sources list
-  protected _loadSources() {
-    // Obtain sources iterator
-    const sources = this._pending!.sources!;
-    this._pending!.loading = true;
+  // Adds the given source to the internal sources array
+  protected _addSource(source: InternalSource<any>) {
+    source[DESTINATION] = this;
+    source.on('error', destinationEmitError);
```

Comment:
This is the second of two behaviors I am not happy with. In particular; errors will only be propogated once we have pulled it from `this._source`. An error may occur before this.